### PR TITLE
Add support for importing/exporting local modules

### PR DESCRIPTION
**Pull request by ianyong**
## Overview

Adds support for importing/exporting local modules for Source 2 onward. This is because our transformation makes use of data structures such as lists and pairs, which are only available in Source 2+.

Part of https://github.com/source-academy/frontend/issues/2176.

## What is a Local Module?

In Source, we make the distinction between the following modules:
1. Source modules, which can be thought of as libraries which can be imported from. There is a separate system for handling these in `js-slang`. The modules that are available can be found in the [`modules` repository](https://github.com/source-academy/modules).
1. Local modules, which are essentially user-written files to allow for multi-file Source programs.

To distinguish between the two, we make use of the `isSourceModule` function:
```ts
/**
 * Returns whether a string is a file path. We define a file
 * path to be any string containing the '/' character.
 *
 * @param value The value of the string.
 */
export const isFilePath = (value: string): boolean => {
  return value.includes('/')
}

/**
 * Returns whether a module name refers to a Source module.
 * We define a Source module name to be any string that is not
 * a file path.
 *
 * Source module import:           `import { x } from "module";`
 * Local (relative) module import: `import { x } from "./module";`
 * Local (absolute) module import: `import { x } from "/dir/dir2/module";`
 *
 * @param moduleName The name of the module.
 */
export const isSourceModule = (moduleName: string): boolean => {
  return !isFilePath(moduleName)
}
```

## Supported ESTree AST Nodes

Given that Source is a subset of JavaScript, the syntax for [imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) and [exports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/export) is the same as JavaScript's. We make use of the [Acorn parser](https://github.com/acornjs/acorn) to parse Source programs into their Abstract Syntax Tree (AST) representation. This particular representation is (mostly) compliant with the [ESTree spec](https://github.com/estree/estree).

As such, we have the following import/export-related AST nodes to work with:
* `ExportNamedDeclaration` - any named exports.
  ```js
  const x = 5;
  function square(x) {
    return x * x;
  }
  export { x as y, square };
  ```
* `ExportDefaultDeclaration` - the default export of a file/module.
  ```js
  function cube = x => x * x * x;
  export default cube;
  ```
* `ExportAllDeclaration` - re-exporting/aggregating values exported from other modules.
  ```js
  export * from "./b.js";
  ```
* `ImportDeclaration` - the parent AST node of all import specifiers.
* `ImportSpecifier` - any named imports.
  ```js
  import { x as y, z } from "./a.js";
  ```
* `ImportDefaultSpecifier` - the default import of a file/module.
  ```js
  import x from "./a.js";
  ```
* `ImportNamespaceSpecifier` - imports all exported values from a module as an object.
  ```js
  import * as x from "./a.js";
  ```

The Source chapters in which they are enabled are:
* 1 - `ImportDeclaration`, `ImportSpecifier`
  * Needed for Source module imports.
* 2 - `ExportNamedDeclaration`
  * For Source 1 through 4, only named imports/exports is enabled. However, the implementation actually supports all AST nodes if possible so that they can be enabled by changing their values in the `syntaxBlacklist`.
* Library parser language (100) - `ExportDefaultDeclaration`, `ImportDefaultSpecifier`
  * Default exports are implemented but not enabled in Source 1 through 4. They can however be used in the library parser language.
* Infinity (cannot be used at all) - `ExportAllDeclaration`, `ImportNamespaceSpecifier`
  * `ExportAllDeclaration` will eventually be implemented, but not enabled in Source 1 through 4. Because of that, it is low priority and will be added at a later date.
  * `ImportNamespaceSpecifier` cannot be supported as Source does not allow dot notation (to access properties on objects).

## How Does it Work?

We handle the importing/exporting of local modules as a preprocessing step. Given a file system (represented as an object of absolute file paths to file contents) and an entrypoint file path, we recursively parse files into their AST form, starting with the entrypoint file. Based on the import declarations in the entrypoint file, we then determine which files to parse next and keep doing this recursively until we have built up an import graph.

The import graph is a directed graph. We run Kahn's algorithm on it to find a topological ordering of nodes (in this case, files/modules). If a valid topological ordering is not found, then the Source program contains a circular dependency and evaluation is stopped with an error.

Otherwise, every file/module involved in the program except the entrypoint file is transformed into a function declaration. This is so that each imported file/module has its own scope. Using the topological ordering, we are able to invoke each of these functions in a manner that lets us pass the result of functions into other functions as required.

Each of these transformed functions returns a pair whose head is the default export or `null` if none exist, and whose tail is a list of pairs that map names to variable/function identifiers. To extract the information required from this structure, we insert the following functions as prelude to Source 2+ programs:

```ts
function __access_named_export__(named_exports, lookup_name) {
  if (is_null(named_exports)) {
    return undefined;
  } else {
    const name = head(head(named_exports));
    const identifier = tail(head(named_exports));
    if (name === lookup_name) {
      return identifier;
    } else {
      return __access_named_export__(tail(named_exports), lookup_name);
    }
  }
}

function __access_export__(exports, lookup_name) {
  if (lookup_name === "default") {
    return head(exports);
  } else {
    const named_exports = tail(exports);
    return __access_named_export__(named_exports, lookup_name);
  }
}
```

Finally, we assemble all of the transformed functions, function invocations, and `__access_export__` calls together to form a single AST. This AST is a valid Source program and the rest of the evaluation is unchanged.

## Example Program

If we have the following Source program consisting of 4 files where `/a.js` is the entrypoint file:
* `/a.js`
  ```js
  import { a as x, b as y } from "./b.js";

  x + y;
  ```
* `/b.js`
  ```js
  import y, { square } from "./c.js";

  const a = square(y);
  const b = 3;
  export { a, b };
  ```
* `/c.js`
  ```js
  import { mysteryFunction } from "./d.js";

  const x = mysteryFunction(5);
  export function square(x) {
    return x * x;
  }
  export default x;
  ```
* `/d.js`
  ```js
  const addTwo = x => x + 2;
  export { addTwo as mysteryFunction };
  ```

The resulting AST of the transformation will be equivalent to the following Source code:
```js
function __$b$dot$js__(___$c$dot$js___) {
  const y = __access_export__(___$c$dot$js___, "default");
  const square = __access_export__(___$c$dot$js___, "square");

  const a = square(y);
  const b = 3;

  return pair(null, list(pair("a", a), pair("b", b)));
}

function __$c$dot$js__(___$d$dot$js___) {
  const mysteryFunction = __access_export__(___$d$dot$js___, "mysteryFunction");

  const x = mysteryFunction(5);
  function square(x) {
    return x * x;
  }

  return pair(x, list(pair("square", square)));
}

function __$d$dot$js__() {
  const addTwo = x => x + 2;

  return pair(null, list(pair("mysteryFunction", addTwo)));
}

const ___$d$dot$js___ = __$d$dot$js__();
const ___$c$dot$js___ = __$c$dot$js__(___$d$dot$js___);
const ___$b$dot$js___ = __$b$dot$js__(___$c$dot$js___);

const x = __access_export__(___$b$dot$js___, "a");
const y = __access_export__(___$b$dot$js___, "b");

x + y;
```

## Current Status

Note that the preprocessing step is currently **not enabled**. This is because the output of the preprocessing step is a transformed AST. However, the Source runner currently takes in the program code as string. It will need to be refactored to operate directly on an AST instead. Initially, we explored the option of transpiling the transformed AST back into code so that this refactoring would not be needed. However doing so results in the location information (column + line number) being lost, meaning that error messages could not properly display the location of the erroneous code.

### CODE REVIEW

**Comment by ianyong** [comment]
@martin-henz Looking at the files with missing coverage as highlighted by coveralls, it seems that the overall coverage decreased due to my use of defensive programming. I replaced quite a few instances of the code where type-casting and the non-null assertion operators were previously used with checks that throw errors instead (such as in https://github.com/source-academy/js-slang/pull/1332/commits/4f9f3a8bf080fc6a2728f07261e76d5d04537526) so as to have better type safety.

image

This is necessary because some invariants (such as `FunctionDeclaration` nodes always having an identifier) are reliant on other parts of the codebase to be true. If one day, we remove the `noFunctionDeclarationWithoutIdentifier` rule (https://github.com/source-academy/js-slang/pull/1332/commits/6ff73b19490239486aba533deaf609c5ccdfeabf), this invariant will be broken.

Unfortunately, these checks will never be false under normal usage (assuming that our invariants are correct), so the only way to get these lines to be covered would be to create test cases that evaluate ASTs that we create by hand.

**Comment by ianyong** [comment]
I will be documenting this feature [here](https://github.com/source-academy/js-slang/wiki/Local-Module-Import-&-Export) while awaiting review (nothing written yet). I expect the review process to take a while due to the scope of this PR.

**Review by zhaojj2209 — APPROVED** [review]
Code looks good, built and tested local copy of js-slang with frontend, frontend behaviour for running single file remains consistent.

File: src/index.ts [hunk 1]
```diff
@@ -284,6 +297,31 @@ export async function runInContext(
   context: Context,
   options: Partial<IOptions> = {}
 ): Promise<Result> {
+  const defaultFilename = 'default'
```
[the remark this answers is not in the document]
Comment (ianyong) [reply src/index.ts:300]: Oops, yes. Thanks for the catch!

**Pull request merged**